hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/HoodieFileGroupReaderBasedParquetFileFormat.scala

```diff
@@ -145,7 +150,7 @@ class HoodieFileGroupReaderBasedParquetFileFormat(tableState: HoodieTableState,
     val fixedPartitionIndexes = fixedPartitionIndexesArr.toSet
 
     // schema that we want fg reader to output to us
-    val requestedSchema = StructType(requiredSchema.fields ++ partitionSchema.fields.filter(f => mandatoryFields.contains(f.name)))
+    val requestedSchema = getRequestedSchema(options, dataSchema, partitionSchema, requiredSchema, mandatoryFields)
```

Comment:
`HoodieFileGroupReaderSchemaHandler#generateRequiredSchema` has the logic to append all necessary fields for merging in MOR which is independent of engines.  Could you check why that's not applied for the bug you discovered?

Also, could we avoid such logic of appending all necessary fields for merging at the Spark layer if possible, to reduce the complexity in this class?